src/components/transport_manager/src/websocket/websocket_transport_adapter.cc

```diff
@@ -0,0 +1,97 @@
+ * this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ * Neither the name of the Ford Motor Company nor the names of its contributors
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "transport_manager/websocket/websocket_transport_adapter.h"
+
+#include <errno.h>
+#include <memory.h>
+#include <signal.h>
+#include <stdio.h>
+
+#include <cstdlib>
+#include <sstream>
+
+#include "transport_manager/websocket/websocket_listener.h"
+#include "utils/logger.h"
+#include "utils/threads/thread_delegate.h"
+
+namespace transport_manager {
+namespace transport_adapter {
+
+CREATE_LOGGERPTR_GLOBAL(logger_, "WebSocketTransportAdapter")
+
+WebSocketTransportAdapter::WebSocketTransportAdapter(
+    resumption::LastState& last_state, const TransportManagerSettings& settings)
+    : TransportAdapterImpl(nullptr,
+                           nullptr,
+                           new WebSocketListener(this, settings),
+                           last_state,
+                           settings) {}
+
+WebSocketTransportAdapter::~WebSocketTransportAdapter() {}
+
+void WebSocketTransportAdapter::TransportConfigUpdated(
+    const TransportConfig& new_config) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  transport_config_ = new_config;
+
+  // call the method of parent class to trigger OnTransportConfigUpdated() for
+  // the listeners
+  TransportAdapterImpl::TransportConfigUpdated(new_config);
+}
+
+TransportConfig WebSocketTransportAdapter::GetTransportConfiguration() const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  return transport_config_;
+}
+
+DeviceType WebSocketTransportAdapter::GetDeviceType() const {
+  return WEBENGINE_WEBSOCKET;
+}
+
+DeviceSptr WebSocketTransportAdapter::AddDevice(DeviceSptr device) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  webengine_device_ = device;
+  return TransportAdapterImpl::AddDevice(webengine_device_);
+}
+
+TransportAdapter::Error WebSocketTransportAdapter::Init() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  if (webengine_device_) {
+    AddDevice(webengine_device_);
```

Comment:
Wondering why this transport was designed to use one webengine device for all connections? As a server role I thought that each incoming connection would count as a device, similar to how the tcp server transport is implemented.

Reply:
@JackLivio The proposal concept differs from TCP devices. You are right that every different TCP connection could be considered as a new device. It is possible to have many TCP devices as they are external ones. The original proposal requires WebEngine device creation before it actually connects. 
`UpdateAppList` for local applications ( which have field `enabled = True`) should be sent to the HMI immediately after HMI sends `SetAppProperties` to SDL core. At this moment SDL can not have any connection from WebEngine device. But UpdateAppList RPC has parameter `HMIApplication` which in its turn has mandatory `deviceInfo` parameter. That's why we have to create WebEngine device before it actually connects. 
Now it is considered that Vehicle can not have many WebEngine devices within one head unit. In order to make it more clear, see the following sequence:

1) OEM store somehow provides information about installed local apps to the HMI
2) HMI sends SetAppProperties request to the SDL core with information about these applications (with `enabled = True`) 
3) SDL sends UpdateAppList to the HMI with enabled local and cloud apps (if any) and related info to the HMI. 
4) HMI shows these application on its screen.
5) Now user can see local (as well as cloud) app list and can launch these apps or one app. 
6) After user push app in the HMI list - HMI can differentiate to trigger local app launch or launch of cloud app. 
7) If HMI discovers that user tries to launch local app - it triggers WebEngine device launch. Webengine device knows which application to start and starts it.
8) Only at this point SDL receives incoming connection request using websocket transport form WebEngine device and accepts it.
The next flow is the same as for regular apps.
SDL receives RAI request and so on...
Current SDL core implementation is based on assumption that on one HMI only one WebEngine device can present. 

If it is required to have many WebEngine devices on one HMI (on one host) - 
It will probably require proposal change. At least conditions how to configure muptiple WebEngine devices, their security, local applications sharing between many webengine devices.

Reply:
I agree with @AByzhynar and the local WebEngine should be treated as a single device running all applications. This allows a single device consent for local WebEngine apps. As a server role each incoming connection should count as an app whereby the remote hostname/address is the device.

Reply:
Ok thank you for the information. I was just looking for some background. I am ok with a single device instance.